flooddrake/timestepper.py

```diff
@@ -0,0 +1,235 @@
+        
+        solve(L==0,w_,nest=False,solver_parameters={'ksp_type':'preonly','pc_type':'lu'})
+        
+        return w_
+    
+    def stepper(self,t_start,t_end,w):
+        
+        """ Timesteps the shallow water equations from t_start to t_end using a 3rd order Runge-Kutta scheme
+        
+        	:param t_start: start time
+        	:type t_start: float
+        	
+        	:param t_end: end time
+        	:type t_end: float
+        	
+        	:param w: Current state vector function
+        
+        """
+        
+        w_old = Function(self.V).assign(w) # used as the original state vector in each RK3 step
+        
+        # raise timestep error
+        if (self.dt.dat.data>t_end):
+            raise ValueError('end time is less than one timestep or timestep is not factor of end time')
+            
+        Nt=int(t_end/self.dt.dat.data)
+        
+        
+        ####################### RUNGE-KUTTA SCHEME ####################
+        
+        if self.mesh.geometric_dimension()==1:
+            
+            # index functions and spaces
+            h, mu= split(w)
+            v_h, v_mu= split(self.V)
+       
+        if self.mesh.geometric_dimension()==2:
+            
+            # index functions and spaces
+            h, mu, mv = split(w)
+            v_h, v_mu, v_mv = split(self.V)
+        
+        v_h, v_mu, v_mv = split(self.V)
+        
+        w = SlopeModification(w,self.V)
+        
+        hout = Function(v_h)
+        hout_file = File("h.pvd")
+        
+        hout.project(h+self.b_)
+        hout_file << hout
+        
+        
+        for i in range(Nt):
+            
+            #
+            G=Function(v_h)
+            G.project(h)
+            A = assemble(G*dx)
+            #
+            
+            w = SlopeModification(w,self.V)
+            
+            w_ = self.__solver_variable(w)
+            
+            w.assign(w_)
+            w = SlopeModification(w,self.V)
+            
+            w_ = self.__solver_variable(w)
+            
+            w.assign((3.0/4.0)*w_old + (1.0/4.0)*w_)
+            w = SlopeModification(w,self.V)
+            
+            w_ = self.__solver_variable(w)
+            
+            w.assign((1.0/3.0)*w_old + (2.0/3.0)*w_)
+            w = SlopeModification(w,self.V)
+            
+            w_old.assign(w)
+            
+            # timstep complete
+            
+            if self.mesh.geometric_dimension()==1:
+                
+                # index functions and spaces
+                h, mu= split(w)
+                v_h, v_mu= split(self.V)
+       
+            if self.mesh.geometric_dimension()==2:
+            
+                # index functions and spaces
+                h, mu, mv = split(w)
+                v_h, v_mu, v_mv = split(self.V)
```

Comment:
This whole implementation is very inefficient since you create and destroy new Functions and symbolic expressions inside the timestepping loop.  It will therefore run very slowly.

I recommend refactoring so that you create the forms and solver in a *setup* phase (plausibly inside `__init__`) and then just use them in the stepper.  You may need to modify things such that (for example) your slopemodification and similar take an input and an output function, so you can reuse both.